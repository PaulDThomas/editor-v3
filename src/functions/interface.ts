// Interfaces
export interface iColourStyles {
  [styleName: string]: React.CSSProperties;
}

export interface iStyleBlock {
  start: number;
  end: number;
  styleName?: string;
}

export interface iColouredLine {
  text: string;
  styles?: iColourStyles;
  styleBlocks?: iStyleBlock[];
}

export enum EditorV3Align {
  'left' = 'left',
  'center' = 'center',
  'decimal' = 'decimal',
  'right' = 'right',
}

export interface AieStyleMap {
  [styleName: string]: { css: React.CSSProperties; aieExclude: string[] };
}
export interface AieStyleExcludeMap {
  [styleName: string]: string[];
}
