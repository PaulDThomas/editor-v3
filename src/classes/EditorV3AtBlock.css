.aiev3 ul.aiev3-at-dropdown-list {
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 1;
  width: max-content;
  max-width: 200px;
  display: block;
  list-style-type: none;
  margin-block-start: 2px;
  margin-block-end: 2px;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
  padding-inline-start: 40px;
  background-color: #fff;
  opacity: 1;
  border: 1px solid #ccc;
  border-radius: 4px;
  padding: 4px;
  font-family:
    -apple-system,
    BlinkMacSystemFont,
    Segoe UI,
    Roboto,
    Oxygen,
    Ubuntu,
    Cantarell,
    Fira Sans,
    Droid Sans,
    Helvetica Neue,
    sans-serif;
  font-size: 12px;
  color: #000;
  font-weight: 400;
}

.aiev3 li.aiev3-at-no-items {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  cursor: not-allowed;
}

.aiev3 li.aiev3-at-item {
  cursor: pointer;
}

.aiev3 li:hover {
  background-color: #f0f0f0;
}
