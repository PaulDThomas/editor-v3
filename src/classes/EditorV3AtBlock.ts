import "./EditorV3AtBlock.css";
import {
  EditorV3TextBlock,
  IEditorV3TextBlock,
  IEditorV3TextBlockOptionalParams,
} from "./EditorV3TextBlock";
import { EditorV3AtListItem, EditorV3RenderProps, EditorV3WordPosition } from "./interface";

export interface IEditorV3AtBlockOptionalParams extends IEditorV3TextBlockOptionalParams {
  atListFunction?: (
    typedString: string,
  ) => Promise<EditorV3AtListItem<{ [key: string]: string }>[]>;
  maxAtListLength?: number;
  atData?: { [key: string]: string };
}

export interface IEditorV3AtBlock extends IEditorV3TextBlock, IEditorV3AtBlockOptionalParams {}

export class EditorV3AtBlock extends EditorV3TextBlock implements IEditorV3AtBlock {
  /**
   * Promise function that will return list of string values to display
   * @param typedString Text entered into the block so far
   * @returns Promise of array of strings
   */
  public atListFunction: (
    typedString: string,
  ) => Promise<EditorV3AtListItem<{ [key: string]: string }>[]> = () =>
    new Promise((resolve) => resolve([]));
  /**
   * Maximum number of items to display in the returned list
   */
  public maxAtListLength = 10;

  public atData: { [key: string]: string } = {};

  get data(): IEditorV3AtBlock {
    const ret: IEditorV3AtBlock = super.data;
    if (Object.keys(this.atData).length > 0) {
      ret.atData = this.atData;
    }
    return ret;
  }

  get wordPositions(): EditorV3WordPosition[] {
    return [
      {
        line: -1,
        startChar: this.lineStartPosition,
        endChar: this.lineStartPosition + this.text.length,
        isLocked: this.isLocked === true,
      },
    ];
  }

  constructor(
    arg: IEditorV3AtBlock | HTMLSpanElement | DocumentFragment,
    forcedParams?: IEditorV3AtBlockOptionalParams,
  ) {
    // Usual text block constructor
    super(arg, forcedParams);
    // Force type
    this.type = "at";
    // Check for at specific information
    if (arg instanceof HTMLSpanElement) {
      this.furtherHtml(arg);
    } else if (arg instanceof DocumentFragment) {
      if (arg.childNodes.length !== 1)
        throw new Error("EditorV3AtBlock:Constructor: DocumentFragment must have 1 child node");
      else if (!(arg.childNodes[0] instanceof HTMLSpanElement))
        throw new Error(
          "EditorV3AtBlock:Constructor: DocumentFragment child node must be HTMLSpanElement",
        );
      this.furtherHtml(arg.childNodes[0] as HTMLSpanElement);
    } else {
      if (arg.atListFunction) this.atListFunction = arg.atListFunction;
      if (arg.maxAtListLength) this.maxAtListLength = arg.maxAtListLength;
      if (arg.atData) {
        this.atData = arg.atData;
      }
    }
    // Force any parameters
    if (forcedParams) {
      if (forcedParams.atListFunction) this.atListFunction = forcedParams.atListFunction;
      if (forcedParams.maxAtListLength) this.maxAtListLength = forcedParams.maxAtListLength;
      if (forcedParams.atData) this.atData = forcedParams.atData;
    }
  }

  private furtherHtml(arg: HTMLSpanElement) {
    // Copy dataset to atFactory
    Object.keys(arg.dataset).forEach((key) => {
      if (!["styleName", "isLocked", "type", "text", "lineStartPosition"].includes(key)) {
        this.atData[key] = arg.dataset[key] as string;
      }
    });
  }

  // Overload html return
  public toHtml(renderProps: EditorV3RenderProps): DocumentFragment {
    // Create holder
    const ret = new DocumentFragment();
    if (renderProps.currentEl) renderProps.currentEl.append(ret);
    const span = document.createElement("span");
    ret.appendChild(span);
    span.classList.add("aiev3-tb", "at-block");
    span.dataset.type = "at";

    // Add Text node
    const textNode = document.createTextNode(
      this.text.replaceAll("\u200c", "").replace(/^ /, "\u00A0").replace(/ $/, "\u00A0"),
    );

    span.appendChild(textNode);
    // Add style
    if (this.style) {
      span.classList.add(`editorv3style-${this.style}`);
      span.dataset.styleName = this.style;
    }

    // Delete any existing dropdown on render
    const editor = renderProps.editableEl?.closest(".aiev3") as HTMLDivElement | null;
    if (editor) {
      const existingDropdowns = editor.querySelectorAll("ul.aiev3-at-dropdown-list");
      existingDropdowns.forEach((dropdown) => dropdown.remove());
    }

    // Add locked status
    if (this.isLocked || !this.isActive) {
      span.classList.add("is-locked");
      span.dataset.isLocked = "true";
      // Add any other data items from the at item
      this.atData &&
        Object.keys(this.atData).forEach((key) => {
          span.dataset[key] = (this.atData as { [key: string]: string })[key] as string;
        });
    } else if (this.isActive) {
      span.classList.add("is-active");

      // Add in dropdown
      const renderDropdown = () => {
        // Double check class list
        if (!span.classList.contains("is-locked")) {
          // Create download
          const dropdownUl = document.createElement("ul");
          // Append to editor
          span.appendChild(dropdownUl);
          // Set up dropdown internals
          dropdownUl.classList.add("aiev3-at-dropdown-list", "skip-read");
          dropdownUl.contentEditable = "false";
          dropdownUl.innerHTML = "<li>Loading...</li>";
          // Add event listener
          dropdownUl.addEventListener("click", (e: MouseEvent) => {
            e.preventDefault();
            e.stopPropagation();
            // Set long timeout to enable stack to clear
            const atItem = e.target;
            if (
              atItem &&
              atItem instanceof HTMLElement &&
              atItem.classList.contains("aiev3-at-item")
            ) {
              // Update span
              const thisText = atItem.textContent;
              span.textContent = atItem.dataset.text ?? thisText;
              span.classList.remove("is-active");
              span.classList.add("is-locked");
              span.dataset.isLocked = "true";
              // Copy any other data from the atItem
              if (atItem.dataset) {
                Object.keys(atItem.dataset)
                  .filter((key) => !["isLocked", "text"].includes(key))
                  .forEach((key) => {
                    span.dataset[key] = atItem.dataset[key];
                  });
              }
              // Remove dropdown
              dropdownUl.remove();
              // Move cursor to end of span
              const range = document.createRange();
              const selection = window.getSelection();
              range.setStartAfter(span);
              range.collapse(true);
              selection?.removeAllRanges();
              selection?.addRange(range);
            }
          });
          // Add event listener to document
          const clickHandler = (e: MouseEvent) => {
            const target = e.target as HTMLElement;
            const dropdownUl = span.querySelector(".aiev3-at-dropdown-list");
            if (dropdownUl && !dropdownUl.contains(target)) {
              dropdownUl.remove();
              span.classList.remove("is-active");
              span.classList.add("is-locked");
              span.dataset.isLocked = "true";
            }
            document.removeEventListener("click", clickHandler);
          };

          document.addEventListener("click", clickHandler);

          // Position dropdown
          const spanRect = span.getBoundingClientRect();
          const offSetRect = span.offsetParent?.getBoundingClientRect();
          if (offSetRect && spanRect) {
            dropdownUl.style.left = `${spanRect.left - offSetRect.left}px`;
            dropdownUl.style.top = `${spanRect.bottom - offSetRect.top}px`;
          }

          // Add items to dropdown when promise resolves
          const atFunction = renderProps.atListFunction ?? this.atListFunction;
          atFunction(this.text)
            .then((resolvedAtList) => {
              dropdownUl.innerHTML = "";
              if (resolvedAtList.length === 0) {
                const noItems = document.createElement("li");
                noItems.classList.add("aiev3-at-no-items");
                noItems.textContent = "No items found";
                dropdownUl.appendChild(noItems);
              } else {
                resolvedAtList.slice(0, this.maxAtListLength).forEach((atItem) => {
                  const atSpan = atItem.listRender ?? document.createElement("li");
                  if (!atSpan.classList.contains("aiev3-at-item"))
                    atSpan.classList.add("aiev3-at-item");
                  if (!atSpan.dataset.text) atSpan.dataset.text = atItem.text;
                  if (atSpan.textContent === "" || !atSpan.textContent)
                    atSpan.textContent = atItem.text;
                  atSpan.dataset.text = atItem.text;
                  // Add in data from atItem
                  if (atItem.data) {
                    Object.keys(atItem.data).forEach((key) => {
                      if (
                        typeof (atItem.data as { [key: string]: string })[
                          key as keyof typeof atItem.data
                        ] === "string"
                      )
                        atSpan.dataset[key] = (atItem.data as { [key: string]: string })[
                          key as keyof typeof atItem.data
                        ] as string;
                    });
                  }
                  dropdownUl.appendChild(atSpan);
                });
                // Add in more items list item
                if (resolvedAtList.length > this.maxAtListLength) {
                  const atSpan = document.createElement("li");
                  atSpan.classList.add("aiev3-more-items");
                  atSpan.textContent = `...${resolvedAtList.length - this.maxAtListLength} more`;
                  dropdownUl.appendChild(atSpan);
                }
              }
            })
            .catch(() => {
              // For when something goes wrong
              dropdownUl.innerHTML = "";
              const errorItem = document.createElement("li");
              errorItem.classList.add("aiev3-at-items-error");
              errorItem.textContent = "Error fetching list";
              errorItem.style.color = "red";
              dropdownUl.appendChild(errorItem);
            });
        }
      };

      // Throttle render
      let renderTimeout: number | null = null;
      if (!renderTimeout) {
        renderTimeout = window.setTimeout(renderDropdown, 100);
      }
    }
    // Add to current element if specified
    if (renderProps.currentEl) renderProps.currentEl.append(ret);
    // Always return
    return ret;
  }
}
