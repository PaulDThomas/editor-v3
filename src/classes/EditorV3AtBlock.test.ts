import { act } from "react";
import { fireEvent } from "@testing-library/react";
import { EditorV3AtBlock } from "./EditorV3AtBlock";
import { EditorV3DropListItem } from "./interface";

describe("EditorV3AtBlock", () => {
  test("should return a DocumentFragment with the correct structure", () => {
    const text = "Hello, world!";
    const style = "bold";
    const block = new EditorV3AtBlock({ text, style });
    expect(block.wordPositions).toEqual([
      {
        line: -1,
        startChar: 0,
        endChar: 13,
        isLocked: false,
      },
    ]);
    block.setActive(true);
    const result = block.toHtml({});

    // Check if the DocumentFragment contains a single span element
    expect(result.childNodes.length).toBe(1);
    expect(result.firstChild).toBeInstanceOf(HTMLSpanElement);

    const span = result.firstChild as HTMLSpanElement;

    // Check if the span element has the correct class names
    expect(span.classList.contains("aiev3-tb")).toBe(true);
    expect(span.classList.contains("at-block")).toBe(true);
    expect(span.classList.contains("is-active")).toBe(true);
    expect(span.classList.contains("is-locked")).toBe(false);
    expect(span.classList.contains(`editorv3style-${style}`)).toBe(true);

    // Check if the span element has the correct content and attributes
    expect(span.textContent).toBe(text);
    expect(span.dataset.isLocked).not.toBeDefined();
    expect(span.dataset.styleName).toBe(style);
  });

  test("should update forcedParams", async () => {
    const text = "Hello, world!";
    const block = new EditorV3AtBlock({
      text,
      maxAtListLength: 15,
      atData: { tree: "larch" },
      atListFunction: () =>
        new Promise<EditorV3DropListItem<{ tree: string }>[]>((resolve) =>
          resolve([{ text: "Birch", data: { tree: "silver birch" } }]),
        ),
    });
    expect(block.data).toEqual({ text, type: "at", atData: { tree: "larch" } });
    expect(await block.atListFunction("")).toEqual([
      { text: "Birch", data: { tree: "silver birch" } },
    ]);
    expect(block.maxAtListLength).toBe(15);
    const updatedBlock = new EditorV3AtBlock(block.data, {
      maxAtListLength: 20,
      atData: { tree: "oak" },
      atListFunction: () =>
        new Promise<EditorV3DropListItem<{ tree: string }>[]>((resolve) =>
          resolve([{ text: "Oak", data: { tree: "oak" } }]),
        ),
    });
    expect(updatedBlock.data).toEqual({ text, type: "at", atData: { tree: "oak" } });
    expect(await updatedBlock.atListFunction("")).toEqual([{ text: "Oak", data: { tree: "oak" } }]);
    expect(updatedBlock.maxAtListLength).toBe(20);
  });

  test("should return a DocumentFragment when isActive with the correct structure", () => {
    const text = "Hello, world!";
    const style = undefined;
    const block = new EditorV3AtBlock({ text, style, isLocked: true });
    const result = block.toHtml({});

    // Check if the DocumentFragment contains a single span element
    expect(result.childNodes.length).toBe(1);
    expect(result.firstChild).toBeInstanceOf(HTMLSpanElement);

    const span = result.firstChild as HTMLSpanElement;

    // Check if the span element has the correct class names
    expect(span.classList.contains("aiev3-tb")).toBe(true);
    expect(span.classList.contains("at-block")).toBe(true);
    expect(span.classList.contains("is-active")).toBe(false);
    expect(span.classList.contains("is-locked")).toBe(true);
    expect(span.classList.toString().includes("editorv3style")).toBe(false);

    // Check if the span element has the correct content and attributes
    expect(span.textContent).toBe("Hello, world!");
    expect(span.dataset.isLocked).toBe("true");
    expect(span.dataset.styleName).not.toBeDefined();
  });
});

describe("should return a DocumentFragment with a dropdown", () => {
  const atListFunction = (typedString: string) =>
    new Promise<EditorV3DropListItem<{ email: string }>[]>((resolve) =>
      window.setTimeout(() => {
        const available: EditorV3DropListItem<{ email: string }>[] = [
          { text: "@one", data: { email: "one" } },
          { text: "@two", data: { email: "two" } },
          { text: "@three", data: { email: "three" } },
          { text: "@four", data: { email: "four" } },
          { text: "@five", data: { email: "five" } },
          { text: "@six", data: { email: "six" } },
          { text: "@seven", data: { email: "seven" } },
          { text: "@eight", data: { email: "eight" } },
          { text: "@nine", data: { email: "nine" } },
          { text: "@ten", data: { email: "ten" } },
          { text: "@eleven", data: { email: "eleven" } },
          { text: "@twelve", data: { email: "twelve" } },
          { text: "@thirteen", data: { email: "thirteen" } },
          { text: "@fourteen", data: { email: "fourteen" } },
          { text: "@fifteen", data: { email: "fifteen" } },
          { text: "@sixteen", data: { email: "sixteen" } },
          { text: "@seventeen", data: { email: "seventeen" } },
          { text: "@eighteen", data: { email: "eighteen" } },
          { text: "@nineteen", data: { email: "nineteen" } },
          { text: "@twenty", data: { email: "twenty" } },
        ];
        resolve(
          available.filter((item) => item.text.toLowerCase().includes(typedString.toLowerCase())),
        );
      }, 1000),
    );

  test("Ensure previous dropdowns are removed", async () => {
    const text = "@";
    const block = new EditorV3AtBlock({ text, atListFunction });
    block.setActive(true);
    const editor = document.createElement("div");
    document.body.appendChild(editor);
    editor.className = "aiev3";
    const editable = document.createElement("div");
    editable.className = "aiev3-editable";
    editable.contentEditable = "true";
    editor.appendChild(editable);
    const line = document.createElement("div");
    line.className = "aiev3-line";
    editable.appendChild(line);

    const oldDropdown = document.createElement("ul");
    oldDropdown.className = "aiev3-dropdown-list";
    editor.appendChild(oldDropdown);
    oldDropdown.innerHTML = "<li class='aiev3-drop-item'>Old item</li>";
    expect(editor.querySelector(".aiev3-dropdown-list")?.textContent).toEqual("Old item");

    await act(async () => {
      block.toHtml({ editableEl: editable, currentEl: line });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
    });
    expect(editor.querySelector(".aiev3-dropdown-list")?.textContent).not.toEqual("Old item");
    expect(document.body.innerHTML).toMatchSnapshot();
    // Check dropdown click
    const items = document.body.querySelectorAll(".aiev3-drop-item");
    expect(items).toBeDefined();
    expect(items.length).toBe(10);
    fireEvent.mouseDown(items![0]);
    await act(async () => jest.runAllTimers());
    expect(editor.querySelector(".aiev3-drop-item")).toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test("Render dropdown for no objects", async () => {
    const text = "@nothing";
    const block = new EditorV3AtBlock({ text, atListFunction });
    block.setActive(true);
    const div = document.createElement("div");

    await act(async () => {
      block.toHtml({ currentEl: div });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
      expect(document.body.querySelector(".aiev3-dropdown-list")).not.toBeNull();
      expect(document.body.querySelector(".aiev3-dropdown-list")!.textContent).toEqual(
        "Loading...",
      );
    });
    expect(document.body.querySelector(".aiev3-dropdown-list")).not.toBeNull();
    expect(document.body.querySelector(".aiev3-dropdown-list")!.textContent).toEqual(
      "No items found",
    );
  });

  test("Render dropdown for all objects", async () => {
    const text = "@";
    const block = new EditorV3AtBlock({ text, atListFunction });
    block.setActive(true);
    const div = document.createElement("div");
    document.body.appendChild(div);

    block.toHtml({ currentEl: div });
    await act(async () => {
      jest.advanceTimersByTime(100);
    });
    // Run pending timer from the function to display the dropdown list
    expect(document.body.querySelector(".aiev3-dropdown-list")).not.toBeNull();
    expect(document.body.querySelector(".aiev3-dropdown-list")!.textContent).toEqual("Loading...");
    await act(async () => {
      jest.runAllTimers();
    });
    const dropDown = document.body.querySelector(".aiev3-dropdown-list");
    expect(dropDown).not.toBeNull();
    expect(dropDown!.textContent).not.toEqual("Loading...");
    expect(document.body.innerHTML).toMatchSnapshot();

    // Check dropdown click
    const items = dropDown?.querySelectorAll("li.aiev3-drop-item");
    expect(items).toBeDefined();
    expect(items?.length).toBe(10);
    const expectedText = items![0].textContent;
    fireEvent.mouseDown(items![0]);
    await act(async () => jest.runAllTimers());
    expect(document.body.querySelector(".aiev3-drop-item")).toBeNull();
    // Check span contents
    const span = div.querySelector(".aiev3-tb.at-block") as HTMLSpanElement;
    expect(span.textContent).toBe(expectedText);
    expect(span.dataset.isLocked).toBe("true");
    expect(span.dataset.email).toBe("one");
    expect(span.className).toBe("aiev3-tb at-block show-dropdown is-locked");
  });

  test("Render dropdown and click off", async () => {
    const text = "@q";
    const block = new EditorV3AtBlock({
      text,
    });
    block.setActive(true);
    const editor = document.createElement("div");
    editor.className = "aiev3";
    const editable = document.createElement("div");
    editable.className = "aiev3-editable";
    editable.contentEditable = "true";
    editor.appendChild(editable);
    const line = document.createElement("div");
    line.className = "aiev3-line";
    editable.appendChild(line);

    await act(async () => {
      block.toHtml({ currentEl: line });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
    });
    // Check outside dropdown click
    fireEvent.mouseDown(document);
    expect(line.querySelector(".aiev3-at-dropdown-list")).toBeNull();
    expect(line.textContent).toBe("@q");
    expect((line.querySelector(".aiev3-tb.at-block") as HTMLSpanElement)?.dataset.isLocked).toBe(
      "true",
    );
    expect(line.innerHTML).toMatchSnapshot();
  });

  test("Render error in list", async () => {
    const text = "@some error";
    const errorCall = jest.fn().mockRejectedValue(new Error("A bad thing happened"));
    const block = new EditorV3AtBlock({
      text,
      atListFunction: errorCall,
    });
    block.setActive(true);
    const div = document.createElement("div");

    await act(async () => {
      block.toHtml({ currentEl: div });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
      expect(document.body.querySelector(".aiev3-dropdown-list")).not.toBeNull();
      expect(document.body.querySelector(".aiev3-dropdown-list")!.textContent).toEqual(
        "Loading...",
      );
      // Resolve the promise in errorCall
      expect(errorCall).toHaveBeenCalled();
    });
    expect(document.body.querySelector(".aiev3-dropdown-list")).not.toBeNull();
    expect(document.body.querySelector(".aiev3-dropdown-list")!.textContent).toEqual(
      "Error fetching list",
    );
  });

  test("Use data and linkRenderer", async () => {
    // Create function
    const atListFunction = async (typedString: string) => {
      const items = Array.from({ length: 26 }, (_, i) => String.fromCharCode(97 + i).repeat(2)).map(
        (letter) => `${letter}`,
      );
      const filter = items.filter((item) =>
        item.toLowerCase().includes(typedString.slice(1).toLowerCase()),
      );
      const ret = filter.map((text, ix) => {
        const retItem: EditorV3DropListItem<{ email: string }> = {
          text,
          data: { email: `${text}@${ix}.com` },
        };
        retItem.listRender = document.createElement("li");
        retItem.listRender.textContent = `George ${text} (${text}@${ix}.com)`;
        return retItem;
      });
      return ret;
    };
    // Create block
    const text = "@";
    const block = new EditorV3AtBlock({ text, atListFunction, maxAtListLength: 20 });
    block.setActive(true);
    const editor = document.createElement("div");
    document.body.appendChild(editor);
    editor.className = "aiev3";
    const editable = document.createElement("div");
    editable.className = "aiev3-editable";
    editable.contentEditable = "true";
    editor.appendChild(editable);
    const line = document.createElement("div");
    line.className = "aiev3-line";
    editable.appendChild(line);

    await act(async () => {
      block.toHtml({ editableEl: editable, currentEl: line });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
    });
    expect(document.body.innerHTML).toMatchSnapshot();

    // Check dropdown click
    const items = document.body.querySelectorAll(".aiev3-drop-item");
    expect(items).toBeDefined();
    expect(items.length).toBe(20);
    fireEvent.mouseDown(items![0]);
    await act(async () => jest.runAllTimers());
    expect(line.querySelector(".aiev3-drop-item")).toBeNull();
    expect(line.textContent).toBe("aa");
    expect(document.body.innerHTML).toMatchSnapshot();

    // Check data
    const readBack = new EditorV3AtBlock(line.children[0] as HTMLSpanElement);
    expect(readBack.data).toEqual({
      text: "aa",
      type: "at",
      isLocked: true,
      style: undefined,
      atData: {
        email: "aa@0.com",
      },
    });

    // Eat your own tail
    expect(new EditorV3AtBlock(readBack.data).data).toEqual(readBack.data);
    expect(new EditorV3AtBlock(readBack.toHtml({})).data).toEqual(readBack.data);
  });
});

describe("EditorV3AtBlock errors", () => {
  test("should throw an error from an empty document fragment", async () => {
    expect(() => {
      new EditorV3AtBlock(document.createDocumentFragment());
    }).toThrow("EditorV3AtBlock:Constructor: DocumentFragment must have 1 child node");
  });
  test("should throw an error from an unknown node type", async () => {
    const frag = new DocumentFragment();
    frag.appendChild(new Text("Bad text!"));
    expect(() => {
      new EditorV3AtBlock(frag);
    }).toThrow("EditorV3AtBlock:Constructor: Child node must be HTMLSpanElement");
  });
});

describe("EditorV3AtBlock markdown output", () => {
  test("should return the correct markdown output", async () => {
    const testBlock = new EditorV3AtBlock({
      text: "@Hello",
      style: "shiny",
      label: "label",
      atData: {
        email: "aa@0.com",
      },
    });
    const result = testBlock.toMarkdown();
    // eslint-disable-next-line quotes
    expect(testBlock.toMarkdown()).toEqual('@[shiny::label::@Hello**{"email":"aa@0.com"}@]');
    const eatOwnTail = new EditorV3AtBlock(result);
    expect(eatOwnTail.data).toEqual(testBlock.data);
  });

  test("should have no data from bad JSON in tyhe markdown", async () => {
    // eslint-disable-next-line quotes
    const testMarkdown = '@[shiny::label::@Hello**{"email":wooo!}@]';
    const testBlock = new EditorV3AtBlock(testMarkdown);
    expect(testBlock.data).toEqual({
      text: "@Hello",
      type: "at",
      style: "shiny",
      label: "label",
    });
  });
});
