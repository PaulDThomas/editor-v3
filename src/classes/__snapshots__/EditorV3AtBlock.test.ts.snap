// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a DocumentFragment with a dropdown Ensure previous dropdowns are removed 1`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb at-block is-active" data-type="at">@<ul style="z-index: 1000; left: 0px; top: 0px;" class="aiev3-dropdown-list skip-read"><li class="aiev3-drop-item" data-text="@one" data-email="one">@one</li><li class="aiev3-drop-item" data-text="@two" data-email="two">@two</li><li class="aiev3-drop-item" data-text="@three" data-email="three">@three</li><li class="aiev3-drop-item" data-text="@four" data-email="four">@four</li><li class="aiev3-drop-item" data-text="@five" data-email="five">@five</li><li class="aiev3-drop-item" data-text="@six" data-email="six">@six</li><li class="aiev3-drop-item" data-text="@seven" data-email="seven">@seven</li><li class="aiev3-drop-item" data-text="@eight" data-email="eight">@eight</li><li class="aiev3-drop-item" data-text="@nine" data-email="nine">@nine</li><li class="aiev3-drop-item" data-text="@ten" data-email="ten">@ten</li><li class="aiev3-more-items">...10 more</li></ul></span></div></div>"`;

exports[`should return a DocumentFragment with a dropdown Ensure previous dropdowns are removed 2`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb at-block is-locked" data-type="at" data-is-locked="true" data-email="one">@one</span></div></div>"`;

exports[`should return a DocumentFragment with a dropdown Render dropdown and click off 1`] = `"<span class="aiev3-tb at-block is-locked" data-type="at" data-is-locked="true">@q</span>"`;

exports[`should return a DocumentFragment with a dropdown Render dropdown for all objects 1`] = `"<span class="aiev3-tb at-block is-active" data-type="at">@<ul style="z-index: 1000; left: 0px; top: 0px;" class="aiev3-dropdown-list skip-read"><li class="aiev3-drop-item" data-text="@one" data-email="one">@one</li><li class="aiev3-drop-item" data-text="@two" data-email="two">@two</li><li class="aiev3-drop-item" data-text="@three" data-email="three">@three</li><li class="aiev3-drop-item" data-text="@four" data-email="four">@four</li><li class="aiev3-drop-item" data-text="@five" data-email="five">@five</li><li class="aiev3-drop-item" data-text="@six" data-email="six">@six</li><li class="aiev3-drop-item" data-text="@seven" data-email="seven">@seven</li><li class="aiev3-drop-item" data-text="@eight" data-email="eight">@eight</li><li class="aiev3-drop-item" data-text="@nine" data-email="nine">@nine</li><li class="aiev3-drop-item" data-text="@ten" data-email="ten">@ten</li><li class="aiev3-more-items">...10 more</li></ul></span>"`;

exports[`should return a DocumentFragment with a dropdown Use data and linkRenderer 1`] = `"<span class="aiev3-tb at-block is-active" data-type="at">@<ul style="z-index: 1000; left: 0px; top: 0px;" class="aiev3-dropdown-list skip-read"><li class="aiev3-drop-item" data-text="aa" data-email="aa@0.com">George aa (aa@0.com)</li><li class="aiev3-drop-item" data-text="bb" data-email="bb@1.com">George bb (bb@1.com)</li><li class="aiev3-drop-item" data-text="cc" data-email="cc@2.com">George cc (cc@2.com)</li><li class="aiev3-drop-item" data-text="dd" data-email="dd@3.com">George dd (dd@3.com)</li><li class="aiev3-drop-item" data-text="ee" data-email="ee@4.com">George ee (ee@4.com)</li><li class="aiev3-drop-item" data-text="ff" data-email="ff@5.com">George ff (ff@5.com)</li><li class="aiev3-drop-item" data-text="gg" data-email="gg@6.com">George gg (gg@6.com)</li><li class="aiev3-drop-item" data-text="hh" data-email="hh@7.com">George hh (hh@7.com)</li><li class="aiev3-drop-item" data-text="ii" data-email="ii@8.com">George ii (ii@8.com)</li><li class="aiev3-drop-item" data-text="jj" data-email="jj@9.com">George jj (jj@9.com)</li><li class="aiev3-drop-item" data-text="kk" data-email="kk@10.com">George kk (kk@10.com)</li><li class="aiev3-drop-item" data-text="ll" data-email="ll@11.com">George ll (ll@11.com)</li><li class="aiev3-drop-item" data-text="mm" data-email="mm@12.com">George mm (mm@12.com)</li><li class="aiev3-drop-item" data-text="nn" data-email="nn@13.com">George nn (nn@13.com)</li><li class="aiev3-drop-item" data-text="oo" data-email="oo@14.com">George oo (oo@14.com)</li><li class="aiev3-drop-item" data-text="pp" data-email="pp@15.com">George pp (pp@15.com)</li><li class="aiev3-drop-item" data-text="qq" data-email="qq@16.com">George qq (qq@16.com)</li><li class="aiev3-drop-item" data-text="rr" data-email="rr@17.com">George rr (rr@17.com)</li><li class="aiev3-drop-item" data-text="ss" data-email="ss@18.com">George ss (ss@18.com)</li><li class="aiev3-drop-item" data-text="tt" data-email="tt@19.com">George tt (tt@19.com)</li><li class="aiev3-more-items">...6 more</li></ul></span>"`;

exports[`should return a DocumentFragment with a dropdown Use data and linkRenderer 2`] = `"<span class="aiev3-tb at-block is-locked" data-type="at" data-is-locked="true" data-email="aa@0.com">aa</span>"`;
